### ISSUE TITLE
Investigate direct support for OpenTelemetry tracing

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#29281]
EF Core currently doesn't support OpenTelemetry, but rather traces via DiagnosticSource.

An OpenTelemetry contrib package exists for instrumenting EF Core to emit activities for command execution (containing a hard-coded translation table for the database name). We should look into providing a more 1st-class support for this.

**Comment by ben-wilson-mews** [comment]
Just adding `System.Diagnostics.Activity` as a keyword for anyone else searching for a similar issue. Thanks for raising this

**Comment by martincostello** [comment]
👋 - I'm one of the maintainers of the OpenTelemetry.Instrumentation.EntityFrameworkCore NuGet package.

A user recently opened this issue, which has opened a can of worms regarding how to generate meaningful traces for non-relational EFCore providers: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/3370

In an ideal world we could deprecate the package and EFCore would have metrics and tracing built-in, but even in the best case scenario we'd have to wait at least year for .NET 11 for that to happen.

In the meantime, is the EFCore team able to provide any guidance on what the best way for us to extend support for non-relational providers would be without needing to do things like special-case CosmosDB etc?

**Comment by roji** [comment]
Hey @martincostello 👋 

EF does have metrics built-in - [see this page](https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/metrics?tabs=windows). Are you providing additionall metrics beyond these that people need? If so, can you provide a bit of details? If these are important, they'd indeed ideally be implemented within EF, as you suggest.

Re tracing... Some higher-level activities are provider-independent, and should work the same way for relational and non-relational; for example, execution of a LINQ query, a call to SaveChanges, etc. Then, lower-level activities such as ExecutedReadItem are very much provider-specific (ReadItem only makes sense on Cosmos), and would indeed require you to special-case that provider and be aware of the specific events that it emits.

I personally don't think it's manageable for you to support all activity across all providers; I'd expect an external instrumentation library like yours to indeed support the first type (EF-wide, provider-independent activities), and not the second. Otherwise you can of course go into having an instrumentation package for each provider, but that seems like a lot...

Ultimately, as you say if there's enough demand for this, it's something we should integrate into EF and its providers (/cc @cincuranet @SamMonoRT).

**Comment by martincostello** [comment]
Thanks @roji.

I wasn't aware of the built-in metrics - I'll read through the documentation more and do some testing to see what there is that matches the [Database Semantic Conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/database-metrics.md) as that's what we would implement on our side. It might mean we can close off https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1330.

For tracing, sounds like we need to review whether we're missing some of the higher-level activities for anything we're missing instead of going to the level of the detail that the user who created the linked issue might have been looking for. Given that the Azure Cosmos SDK has lower-level traces anyway, that's probably the better avenue for detailed CosmosDB support there.

> Otherwise you can of course go into having an instrumentation package for each provider, but that seems like a lot...

Yeah, I don't think we want to go down that route 😄

---

Just to confirm, [these events](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/ffdbf763c64cbdb00536e50c1cbde4b87bd7a586/src/OpenTelemetry.Instrumentation.EntityFrameworkCore/Implementation/EntityFrameworkDiagnosticListener.cs#L16-L19) would all be considered provider-independent?

- Microsoft.EntityFrameworkCore.Database.Command.CommandCreated
- Microsoft.EntityFrameworkCore.Database.Command.CommandExecuting
- Microsoft.EntityFrameworkCore.Database.Command.CommandExecuted
- Microsoft.EntityFrameworkCore.Database.Command.CommandError

**Comment by roji** [comment]
> Just to confirm, [these events](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/ffdbf763c64cbdb00536e50c1cbde4b87bd7a586/src/OpenTelemetry.Instrumentation.EntityFrameworkCore/Implementation/EntityFrameworkDiagnosticListener.cs#L16-L19) would all be considered provider-independent?

Not quite - these events are low-level relational events ("command" here refers to ADO.NET's DbCommand). So they apply to all SQL databases (PG, SQL Server...) but you won't see them when targeting non-relational databases like Cosmos.

There's one more important thing I left out from my response above... In general, I've felt that the lower-level events shouldn't be traced/observed via EF instrumentation, but rather by instrumentation at the lower-level driver. In other words, if someone wants low-level Cosmos tracing events (e.g. ExecutedReadItem), they should be getting that from the Cosmos SDK (or an instrumentation package of that), and not from EF (or from the EF Cosmos provider). Similarly, for PostgreSQL, Npgsql (the low-level driver) already emits OTel tracing events natively, and SqlClient has an instrumentation package that does the same.

Basically, EF has no additional value here - it's just a layer on top of low-level driver, and good tracing should exist at the low-level driver (natively or via external instrumentation) regardless of whether EF is being used or not. Also, each driver has the best knowledge of what should be traced/logged/reporting in metrics in the based way - this really isn't an EF thing.

So to summarize, I'd expect EF to only emit metrics and tracing for higher-level EF events (stuff in CoreEventId, e.g. SaveChangedCompleted), and leave the lower-level things (CommandExecuted) for handling by their respective drivers - totally out of EF.

(FWIW, I believe that for most users the lower-level things are probably a bit more interesting, i.e. how long an individual SQL (or NoSQL) statement took to execute. EF-level events do have value: a single EF LINQ query can be a split query, requiring multiple roundtrips/commands, and full tracing at both levels would show that. The same applies fo SaveChanges, which can send multiple commands to the database. But I'm not sure most users really need to care about the high-level EF events in most cases)

**Comment by roji** [comment]
Am noting that EF is currently lacking a simple QueryExecutionStarting/QueryExecutionCompleted event pair... Opened #37111 to track.

**Comment by Darlan82** [comment]
@roji 
I think you're mixing up the concepts a bit.
The Driver logs and EF logs belong to EF. In some cases, they are redundant, but not always. EF telemetry is at a higher level, as mentioned in the "SaveChanged" case, and in that case, it has to show the total duration of "SaveChanged" regardless of how many commands were given to the driver.

**Comment by roji** [comment]
> The Driver logs and EF logs belong to EF.

Why do you say that the "driver logs" belong to EF (and what exactly do you mean by that? I believe that ideally, each lower-level drivers/SDK emits its own rich telemetry, including tracing data that's specific to that database - that shouldn't be a general EF responsibility.

**Comment by martincostello** [comment]
From https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/3370#issuecomment-3585130995, we'd be willing to help move this forward for .NET 11(?).

It seems that this issue never quite meets the bar for the work to happen within the EFCore team (given the issue is over 3 years old at this point), so needs some extra resource to get it moving.

If the EFCore team could provide at least some more concrete design details on how they would like this to be implemented, then I'd be willing to start looking at doing the actual implementation.

**Comment by roji** [comment]
@martincostello thanks for offering to help, but I'm still trying to understand what exactly would be the value of implementing something like this in EF... Given that there's already ample command instrumentation at the lower level (SqlClient, Npgsql, Cosmos SDK), why isn't that sufficient? What would users get from using the EF-level command instrumentation that they can't already get today from their lower-level driver/SDK?

**Comment by Trivivium** [comment]
To me there are two areas I think EF could help improve telemetry:

1) Set attributes such as `db.collection.name` with the table name(s) affected by a query, or set the `db.operation.name` with `SELECT`, `INSERT`, etc.

2) Provide better options to enrich `INSERT`, `UPDATE`, or `DELETE` statements caused by the `DbSet<T>.{Add,Update,Delete}` methods. For queries, I can use `TagWith(string)`, and then use the `SqlClient` instrumentation to look for the SQL comment during enrichment. Preferably, I would be able to add custom attributes when defining the Linq query, or close to the execution of the Linq query.

Granted, all this information could be parsed from the SQL statement in the SqlClient instrumentation. However, it is cumbersome and error-prone. While I have only surface-level understanding of the EF execution pipeline, I assume EF has access to this information when constructing the SQL query to execute. 

Although I haven't had the chance to use CosmosDB yet, I imagine both the `db.collection.name` and `db.operation.name` attributes map to sensible concepts too.

**Comment by Darlan82** [comment]
@roji 
The goal in my project was to log the instrumentation data on the OLT, however, there are several if statements that prevent capturing the instrumentation data.
There are problems with the combination of EF, Driver, Aspire, and OLT.

**Comment by roji** [comment]
@Trivivium these are good points.

> Set attributes such as db.collection.name with the table name(s) affected by a query, or set the db.operation.name with SELECT, INSERT, etc.

That's indeed true - EF works at a higher level, and can therefore know certain things which the lower-level drivers cannot (without parsing SQL, which they should not do).

I'll note that the opposite also exists: the tracing semantic semantic conventions for PostgreSQL ([link](https://opentelemetry.io/docs/specs/semconv/database/postgresql/)) mandates that `db.namespace` should contain the first SEARCH_PATH component, which a lower-level driver (Npgsql) is likely to know, but not necessarily EF. Outside of the official specs, a lower-level driver can also report things like bytes sent/received and other lower-level things which EF won't be able to know.

Bottom line, EF is better-suited for knowing certain things (operation, table name...), while the lower-level driver is better-suited for others... I agree that the operation/table name do provide a reason to have telemetry implemented at the EF level, but I'm not quite sure it's worth the duplication of effort given that there's already telemetry in the lower-level drivers (or instrumentation packages that provide it)...

> Preferably, I would be able to add custom attributes when defining the Linq query, or close to the execution of the Linq query.

That's an interesting idea as well.

**Comment by roji** [comment]
Another advantage of tracing at the EF level (@vonzshik made me think of this) is that EF is able to properly sanitize inlined parameters ([link](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-10.0/whatsnew#redact-inlined-constants-from-logging-by-default)), whereas a lower-level driver only has the SQL and cannot do any redaction/sanitization (without parsing).

**Comment by martincostello** [comment]
+1 to the above points.

We do parse queries in the OpenTelemetry instrumentation to do things like sanitization and other things to set [span attributes for the database Semantic Conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/db/database-spans.md), but in the case of EFCore we're basically reparsing for a lot of information that EF would already have knowledge of.

I think a native implementation would have fewer moving parts overall, and we could then potentially remove/simplify a bunch of code in the EFCore instrumentation. It would also naturally be in lock-step with the implementation in EFCore itself over time.

**Comment by roji** [comment]
@martincostello just noting that the [specs](https://opentelemetry.io/docs/specs/semconv/database/database-spans/) explicitly forbid parsing e.g. `db.operation` or `db.collection_name` from the SQL (`The collection name SHOULD NOT be extracted from db.query.text`).

I'm also very skeptical about reliably sanitizing SQL of potentially sensitive literals after-the-fact - this requires quite a deep level of SQL parsing that I'd highly advise against (I'm not sure what you're currently doing there).

All these are indeed good reasons for implementing tracing within EF rather than at a lower-level, where we only get a finalized, already-baked SQL.

---

Otherwise, let me have a quick discussion with the team to make sure we're all on the same page. Beyond that, here are some design/implementation thoughts:

* We'd have an ActivitySource specific for Microsoft.EntityFramework.Relational (which would work for all relational providers). Cosmos and other non-relational providers would have their own, separate ActivitySource (e.g. Microsoft.EntityFramework.Cosmos).
* At the very end of query postprocessing, when query compilation is almost finished ([code](https://github.com/roji/efcore/blob/285ef30465e212bbc60bad58aad9f8a275a0b736/src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs)), we'd add a step which runs only if the ActivitySource is enabled, and extract out relevant information from the query tree (operation name, table being queried...).
    * Note that some SQL changes are still made after this point, especially around nullability (in [RelationalParameterBasedSqlProcessor](https://github.com/roji/efcore/blob/285ef30465e212bbc60bad58aad9f8a275a0b736/src/EFCore.Relational/Query/RelationalParameterBasedSqlProcessor.cs)). However, I don't think these should should have an effect on the observability info extracted from the query.
* This bundle of extracted observability information (RelationalObservabilityInfo?) would be stored on [RelationalQueryCompilationContext](https://github.com/roji/efcore/blob/285ef30465e212bbc60bad58aad9f8a275a0b736/src/EFCore.Relational/Query/RelationalQueryCompilationContext.cs), and then passed to the various "QueryingEnumerable" implementations ([code](https://github.com/roji/efcore/blob/285ef30465e212bbc60bad58aad9f8a275a0b736/src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.cs)), which is where queries actually get executed. These would then start and end the Activity, populating it with tags from that bundle whenever a query is executed (e.g. see [SingleQueryingEnumerable](https://github.com/roji/efcore/blob/285ef30465e212bbc60bad58aad9f8a275a0b736/src/EFCore.Relational/Query/Internal/SingleQueryingEnumerable.cs)).
* All that would take care of querying, but the update pipeline would need to be modified as well. This is likely to be considerably easier.
* I'd keep **metrics** firmly in mind when implementing ([specs](https://opentelemetry.io/docs/specs/semconv/database/database-metrics/)), so that we can reuse the same infrastructure to collect and report metrics information as well.

This is all unlikely to be a trivial task, and requires knowing quite a bit about EF internals. However, if you feel like tackling this, let us know (but let me confirm with the team first).

**Comment by martincostello** [comment]
> just noting that the [specs](https://opentelemetry.io/docs/specs/semconv/database/database-spans/) explicitly forbid parsing e.g. `db.operation` or `db.collection_name` from the SQL (`The collection name SHOULD NOT be extracted from db.query.text`).

That's not _quite_ what it says. I think the salient bit is the part I've highlighted in **bold** below:

> The collection name SHOULD NOT be extracted from db.query.text, **when the database system supports query text with multiple collections in non-batch operations.**

[Sanitization](https://opentelemetry.io/docs/specs/semconv/database/database-spans/#sanitization-of-dbquerytext) was implemented for the query summary capture in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2446. It's since had some optimizations (https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/3116) and refactoring for use in EFCore instrumentation too (https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/3022). I'm sure there's room for further improvement.

We're using [this milestone](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/milestone/6) to track work towards stabilising the SqlClient instrumentation library, but given the conversation here we're likely going to leave the EFCore instrumentation at beta quality indefinitely if this work to add native EFCore instrumentation goes ahead.

---

Will keep an eye on this issue to see what the discussion outcome is.

**Comment by roji** [comment]
> That's not quite what it says. I think the salient bit is the part I've highlighted in bold below

Apologies, you're right - an earlier version of the spec was explicit about implementations not parsing/extracting stuff from SQL only for the purpose of tracing; the current spec is indeed different.

Though to be very technical about it, any SQL database "supports query text with multiple collections in non-batch operations", in the sense that any SQL database allows joins. Similarly, SQL operations can support multiple operation types in a single statement (e.g. UPDATE but with a SELECT CTE as input, though that's far more edge-casey than joins). So all that could be understood to indicate that for SQL database these values shouldn't be parsed out; though I admit I don't see the sense of not doing so when there only is one table and operation in the SQL statement.

> I'm sure there's room for further improvement.

Reliably parsing SQL is extremely difficult... I can see various issues in a casual look: for example, table and column names may include a single-quote, and that look like it would be identified as a literal to be sanitized in the current code. To compound the problem, different databases differ here: SQL Server allows quoting table/column names via brackets (so `[Table'WithSingleQuote]`), but MySQL uses backticks, etc. And that's just one problem.

I guess a best-effort sort of approach could be OK here (it generally works but sometimes doesn't :)), though given this is security-related, I'm not sure I'd personally want to deliver something that sometimes works.

> We're using [this milestone](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/milestone/6) to track work towards stabilising the SqlClient instrumentation library, but given the conversation here we're likely going to leave the EFCore instrumentation at beta quality indefinitely if this work to add native EFCore instrumentation goes ahead.

OK, makes sense. FWIW I think it indeed makes sense for EF to have native tracing support for the reasons discussed above - but that indeed doesn't mean we'll have the time/resources to actually do it anytime soon. I'll keep you up to date.

Another point is that even at the SqlClient level, ideally tracing would be provided in-the-box rather than via an external instrumentation library... Though if the EF-level instrumentation is implemented (as above), that would remove much of the need for SqlClient-level instrumentation (all EF users will be able to just use the EF-level instrumentation instead, leaving only non-EF users).

**Comment by roji** [comment]
We think this would be a good addition to EF:

* EF can emit richer tracing data that lower-level drivers cannot (e.g. table names, SELECT vs. UPDATE...).
* We know that built-in OTel tracing in the various drivers isn't very widespread; ideally all would have this, but the reality is that many drivers lack it (e.g. Microsoft.Data.Sqlite).

Putting on the backlog for now.

**Comment by Trivivium** [comment]
To add an additional consideration, it would be awesome if EF Core could add tracing to split-queries too. Today, multiple traces with the similar attributes are emitted for each query. Granted I haven't explored the latest OTel instrumentation package that parses the SQL query to see if it fixes this. But it comes with the trade-offs mentioned earlier.

I think tracing should be considered too, if batching of split-queries #10878 is implemented.
